llvm/include/llvm/IR/DebugLoc.h

```diff
@@ -22,6 +23,67 @@ namespace llvm {
   class LLVMContext;
   class raw_ostream;
   class DILocation;
+  class Function;
+
+#if ENABLE_DEBUGLOC_COVERAGE_TRACKING
+  // Used to represent different "kinds" of DebugLoc, expressing that a DebugLoc
+  // is either ordinary, containing a valid DILocation, or otherwise describing
+  // the reason why the DebugLoc does not contain a valid DILocation.
+  enum class DebugLocKind : uint8_t {
+    // DebugLoc is expected to contain a valid DILocation.
+    Normal,
+    // DebugLoc intentionally does not have a valid DILocation; may be for a
+    // compiler-generated instruction, or an explicitly dropped location.
+    LineZero,
```

Comment:
Hmm, I'd really prefer the name of this case to clearly state the "intentional" aspect. Perhaps `IntentionallyZero`...?

Another approach might be to have separate enum values for the different reasons, e.g. `CompilerGenerated`, `ExplicitlyDropped`, and whatever the others might be. Stating these cases very clearly, not just in comments, but also via enum value would appear to be a very valuable capturing of intent to me. (However, maybe it is too much trouble to do so...? Please say that if so.)

Reply:
Yeah, I didn't have a good name thought up for this - IntentionallyZero might be good, though even the word "Intentional" I'm not 100% on. Long run, the goal is to split this category into different subcategories as you suggested, but only at the point where there is some meaning to separating them - until then I felt it would be premature to try to assign categories, since it's not fully clear what they would be.

Reply:
Looking at it another way, I suppose my point is `LineZero` doesn't convey whether it's on purpose, an accident, or some undetermined mix of the two. But really, this case is trying to capture cases where explicit choices were made, so the name should reflect that somehow.

Reply:
Decided to tackle this by just splitting this straight into two categories - `CompilerGenerated` and `Dropped` - that each more specifically describe the reason for the empty DebugLoc. This more closely aligns with the original plans, which I'd veered off from because there wasn't any technical purpose to distinguishing the two cases, but I think having clear and meaningful names is enough reason!

Reply:
Thanks, this new version looks great to me! 😄